tests/src/test/java/alluxio/master/journal/ufs/UfsJournalLogWriterTest.java

```diff
@@ -178,6 +197,229 @@ public void writeJournalEntryRotate() throws Exception {
     }
   }
 
+  /**
+   * Tests that (1) {@link UfsJournalLogWriter#mJournalOutputStream} is reset when an exception
+   * is thrown, and (2) the {@link UfsJournalLogWriter} recovers during the next write. It should
+   * write out all entries, including the one that initially failed.
+   */
+  @Test
+  public void recoverFromUfsFailure() throws Exception {
+    long startSN = 0x10;
+    UfsJournalLogWriter writer = new UfsJournalLogWriter(mJournal, startSN);
+    final int numberOfEntriesWrittenBeforeFailure = 10;
+    long nextSN = writeJournalEntries(writer, startSN, numberOfEntriesWrittenBeforeFailure);
+    // Create a mock DataOutputStream.
+    DataOutputStream badOut = createMockDataOutputStream(writer);
+    // Specify the behavior of badOut so that it throws an IOException containing
+    // "injected I/O error" when its `write` method is invoked with any arguments
+    // matching (byte[], int, int).
+    Mockito.doThrow(new IOException(INJECTED_IO_ERROR_MESSAGE)).when(badOut)
+        .write(Mockito.any(byte[].class), Mockito.anyInt(), Mockito.anyInt());
+    tryWriteAndExpectToFail(writer, nextSN);
+    // Write another entry. UfsJournalLogWriter will perform recovery first.
+    writer.write(newEntry(nextSN));
+    nextSN++;
+    writer.close();
+
+    checkJournalEntries(startSN, nextSN);
+  }
+
+  /**
+   * Tests that {@link UfsJournalLogWriter#flush()} can recover after Ufs failure.
+   *
+   * This test has two steps.
+   * 1. Write several journal entries, flush and close the journal. This will create a complete
+   *    journal file, i.e. <startSN>-<endSN>.
+   * 2. Write another journal entry, do NOT flush it, and inject an I/O error.
+   * 3. Attempt to write another journal entry, which is expected to fail. After the failure,
+   *    the newly created incomplete journal file may or may not have valid content.
+   * 4. At this point, the UFS does not guarantee the consistency of the incomplete journal. To
+   *    test how {@link UfsJournalLogWriter} recovers from Ufs failure before
+   *    {@link UfsJournalLogWriter#flush()}, we simply delete the possibly created incomplete
+   *    journal.
```

Comment:
I don't think we need this explanation - the inline comments already describe the steps of the test.

Reply:
Done.